### ISSUE TITLE
[BUG] Call counter goes below zero on shared dialog replication

### ISSUE THREAD

**Issue by Ozzyboshi** [OpenSIPS/opensips#1937]
### OpenSIPS version

version: opensips 2.4.6 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on 10:33:48 Jan  7 2020 with gcc 6.3.0

### Description
I have trouble using dialog module on top of clusterer.
In my test environment I have 2 Opensips 2.4.6 nodes (let's call them Opensips1 and Opensips2) with Clusterer module + Dialog module + Cachedb Mongodb replication.
Dialogs are replicated with proto bin clusterer method.

Opensips 1 opensips.cfg file:
```
#### CLUSTERER module
loadmodule "clusterer.so"
modparam("clusterer", "db_url", "mysql://mysqllogin:mysqlpass@mysqlip/opensips")
modparam("clusterer", "db_mode", 1)
modparam("clusterer", "current_id", 1)

#### DIALOG module
loadmodule "dialog.so"
modparam("dialog", "dlg_match_mode", 1)
modparam("dialog", "default_timeout", 30)
modparam("dialog", "db_mode", 0)
modparam("dialog", "db_url","mysql://mysqllogin:mysqlpass@mysqlip/opensips") 
modparam("dialog", "profiles_with_value", "caller/s")
modparam("dialog","cachedb_url","mongodb://#mongodbreplicasetips#/opensipsDB.dialog?replicaSet=opensips")
modparam("dialog", "dialog_replication_cluster", 1)
modparam("dialog", "dlg_sharing_tag", "SPS01=active")
```

Opensips 2 opensips.cfg file:
```
#### CLUSTERER module
loadmodule "clusterer.so"
modparam("clusterer", "db_url", "mysql://mysqllogin:mysqlpass@mysqlip/opensips")
modparam("clusterer", "db_mode", 1)
modparam("clusterer", "current_id", 2)

#### DIALOG module
loadmodule "dialog.so"
modparam("dialog", "dlg_match_mode", 1)
modparam("dialog", "default_timeout", 30)
modparam("dialog", "db_mode", 0)
modparam("dialog", "db_url","mysql://mysqllogin:mysqlpass@mysqlip/opensips") 
modparam("dialog", "profiles_with_value", "caller/s")
modparam("dialog","cachedb_url","mongodb://#mongodbreplicasetips#/opensipsDB.dialog?replicaSet=opensips")
modparam("dialog", "dialog_replication_cluster", 1)
modparam("dialog", "dlg_sharing_tag", "SPS02=active")
```

Later in both opensips.cfg, when a call is established a dialog is created with its own sharing tag...

```
# -----------------------------
#       Create Dialog
# -----------------------------
# create dialog with timeout
if ( !create_dialog("B") ) {
     send_reply("500","Internal Server Error");
      exit;
}
                        
 set_dlg_sharing_tag(#HERE PUT dlg_sharing_tag NAME#);

```

Let's assume a call is processed by Opensips 1 (the call is forwarded through a Load Balancer), a dialog is created, the call counter gets incremented by 1 on Mongodb (on opensipsDB.dialog collection),  then the dialog is replicated to Opensips 2 via proto bin.
I verified that this actually happens issuing 
```
opensipsctl fifo dlg_list_ctx
```
on both nodes and I can see the same results.

When the call comes to an end, Opensips1 gets a SIP Bye Message (my load balancer uses persistent callid and the BYE is always received by the same Opensips who initiated the sip dialog) and the dialog list gets correctly cleared on both nodes.
The call counter on the mongodb also get decremented by 1 (correct) and returns to 0.

What if the call takes more than 30 seconds?
In this case, the default_timeout previously set, triggers a timeout function.
I think it's dlg_ontimeout() in dlg_handlers.c (modules/dialog folder).
Then dlg_end_dlg() is called which in turn calls val_bye_event().
Now, val_bye_event() is a function in dlf_req_within.c, it checks if the state of the dialog has been updated to DLG_STATE_DELETED (this is our case) and then calls

```
remove_dlg_prof_table(dlg,0,0);
```
The second argument is the replication_tag, in other words it seems that if val_bye_event() is called after a timeout, it always passes 0 as a replication_tag.

You can see it at 
https://github.com/OpenSIPS/opensips/blob/d7849e457f78b20696eb0c90c08bbdc90e314aee/modules/dialog/dlg_req_within.c#L239

The replication flag is now forwarded to destroy_linker() in dlg_profile.c via remove_dlg_prof_table(), see 
https://github.com/OpenSIPS/opensips/blob/d7849e457f78b20696eb0c90c08bbdc90e314aee/modules/dialog/dlg_profile.c#L751

Finally, destroy_linker() decrements the call counter in the mongodb on BOTH nodes, becase all of them will have replication_tag equals false.
Here
https://github.com/OpenSIPS/opensips/blob/d7849e457f78b20696eb0c90c08bbdc90e314aee/modules/dialog/dlg_profile.c#L701
we can notice that if replication_tag is false then the condition is true and cdbf.sub (the function that actually decrements the counter in my mongodb instance) is called.

For now I patched my code with something like:

In destroy_linker()
```
int is_backup_state = get_shtag_state(dlg) == SHTAG_STATE_BACKUP;
if (!is_backup_state && cdbf.sub(cdbc, &dlg_prof_val_buf, 1,
                                        profile_timeout, NULL) < 0) {
                                LM_ERR("cannot remove profile from CacheDB\n");
                                return;
                        }

```

In other words, before performing cdbf.sub() I always check if the share tag is in backup state and only if not I perform the mongodb call counter update.
However I think there are better ways do patch this thing, that is why I am asking for your help, first of all I doubt that setting the replication tag always to zero is the right thing to do on timeout events.

For now I'd like to know if you (maintainers of Opensips) acknoledge this issue and if you can produce a better patch. The goal here is just not having negative call counter because that is no-sense.

Thanks in advance for your help and sorry for the long report.

### OS/environment information
 - OS: Debian 9
 - OpenSIPS installation: latest git sources on 2.4.6 branch

**Comment by razvancrainea** [comment]
@Ozzyboshi most likely the problem appears from the fact that you are using different dialog sharing tags (`SPS01` and `SPS02`) on your two instances, and you do not assign any value to the "remote" tag.
What happens is when the timeout triggers, both OpenSIPS instances check if the dialog sharing tag assigned to the dialog, `SPS01`, is active for that specific instance. The first instance will notice it `active`, and decrement the counter. The problem is on the second node, which is not aware of the `SPS01` tag, so it does not consider it as `backup`, but rather as `unknown`, resulting in a second decrement of the counter.
If you are planning to share dialog between two instances, make sure you also use the same dialog sharing tag.

There is however an open discussion here - should we consider an `unknown` tag as `backup`? If so, that would sort out your issue. However, that would break things of there is a tag that comes "out of nowhere" (no one sets an active state for it). So I am not 100% sure how to do this right now.

**Comment by Ozzyboshi** [comment]
@razvancrainea 
Thanks for your reply.
In other words you are suggesting to declare explicitly each sharing tag on each node conf file.
In my example

Node1:
```
modparam("dialog", "dlg_sharing_tag", "SPS01=active")
modparam("dialog", "dlg_sharing_tag", "SPS02=backup")
```

Node2:
```
modparam("dialog", "dlg_sharing_tag", "SPS02=active")
modparam("dialog", "dlg_sharing_tag", "SPS01=backup")
```

I am pretty sure I've already tried this but I can analyze this situation again if you state that is the way to go.

**Comment by razvancrainea** [comment]
Actually, @rvlad-patrascu pointed out that this is the desired behavior - if an instance does not know a tag, it will consider it as backup. It seems I was a bit distracted by the different tags you used, and missed that.
The problem is indeed, as you pointed out, that the `destroy_linker` does not actually check if the node is `active`. @rvlad-patrascu is going to push a fix for this soon.

**Comment by Ozzyboshi** [comment]
@rvlad-patrascu thx for the fix, I will try it as soon as possible
